### PR TITLE
feat: synchronous validators

### PR DESCRIPTION
**Pull request by HendrikThePendric**
This adds a bunch of validators to `ui-forms`. Some notes:
- Not everything can be validated with 100% certainty. In these cases I've focussed mostly on not producing any false positives (i.e. valid inputs being classified as invalid). Rejecting invalid inputs correctly is less important than that.
- Apart from the `required` validator, all validators allow empty values. This makes usage quite flexible. Just combine any validator with required, to enforce the filed to be filled (i.e. `const requiredEmail = composeValidators(required, email)`
- For `email`, `internationalPhone`, and `url` I have tried to provide comprehensive lists of valid and invalid values, so that edge case are well tested. However, I don't feel that confident about the list of test cases I am using for `internationalPhone`. I think quite a few are US numbers.
- Interestingly I found that our `email` regex would allow `あいうえお@example.com`. After some research I found out that there is something called [international email](https://en.wikipedia.org/wiki/International_email) and if our system supports that I think this record should be moved to the list of valid email addresses.
- You will also find a few changes to the integration tests in this PR. That's simply because they broke when I made some changes to the validators that were being used there.
- Finally, you might not find some validators that you were expecting. For example, something to check if the input was in upper or lower case, or whether an input is a valid date. I've omitted these for now, because I am hoping/expecting that these type of cases can be covered by formatters/parsers (i.e. upper/lower-case) or dedicated components (i.e. a DatePicker component should produce valid dates anyway).

Looking forward to getting input on this.

### CODE REVIEW

File: src/validators/alphaNumeric.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import i18n from '@dhis2/d2-i18n'
+import { isEmpty, isString } from './helpers/index.js'
+
+const ALPHA_NUMERIC_PATTERN = /^[a-z0-9]*$/i
```
Comment (Mohammer5) [inline src/validators/alphaNumeric.js]: I think this should allow spaces. I'd expect "This is a text" to be a valid alpha numeric value. If you agree, please add a test value to the test

File: src/validators/createEqualTo.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+const createEqualTo = (key, description) => {
+    requireArgument(key, 'string')
+
+    const errorMessage = i18n.t(
+        'Please make sure the value of this input matches the value in "{{otherField}}".',
+        { otherField: description || key }
+    )
+
+    return (value, allValues) =>
+        isEmpty(value) || isEmpty(allValues[key]) || value === allValues[key]
```
Comment (Mohammer5) [inline src/validators/createEqualTo.js]: We could change this line to just `value === allValues[key]`. If both values are empty, this will return undefined. If the original value is not empty, this should throw an error. I think this validator implies that the value of its field is required when the other value is not empty, otherwise this wouldn't make much sense.

File: src/validators/createMaxCharacterLength.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import i18n from '@dhis2/d2-i18n'
+import { createCharacterLengthRange } from './createCharacterLengthRange'
+
+const createMaxCharacterLength = upperBound =>
+    createCharacterLengthRange(
+        -Infinity,
```
Comment (Mohammer5) [inline src/validators/createMaxCharacterLength.js]: Why this and not 0?

File: src/validators/internationalPhoneNumber.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * - The last part is the subscriber number (SN)
+ * - The NDC and SN together are collectively called the national (significant) number
+ *
+ * IMPLEMENTATION ADVICE
+ * Two important things to note: First of all, in the international E.164 notation a
+ * leading ‘0’ is removed. The UK mobile phone number ‘07911 123456’ in international
+ * format is ‘+44 7911 123456’, so without the first zero. Secondly in the E.164 notation
+ * all spaces, dashes [‘-‘] and parentheses [ ‘(‘ and ‘)’] are removed, besides the
+ * leading ‘+’ all characters should be numeric.
+ */
```
Comment (Mohammer5) [inline src/validators/internationalPhoneNumber.js:28]: So the current implementation would return undefined for this: `+49 0165 123456`, although it doesn't match the E.164 standard, so this validator does not implement the E.164 correctly.

I think a validator is the wrong solution to this problem, if we want to have #.164 compliant phone number. We actually need an international phone number input component, which has an input or select for the country code and a second input for the number which trims leading zeros and special chars and then checks for a max length of 12.

File: src/validators/password.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ * Tests if a given password is compliant with the password restrictions.
+ * This function checks all restrictions below, but returns when the first violation was found:
+ * - At least 8 characters
+ * - No more than 34 characters
+ * - Contains at least 1 lowercase character
+ * - Contains at least 1 UPPERCASE character
+ * - Contains at least 1 number
+ * - Contains at least 1 special character
+ */
+const password = value => {
```
Comment (Mohammer5) [inline src/validators/password.js]: I don't like these static checks here if we call this one just `password`, because it's very dhis2 specific. `ui-core` is meant to work outside of the dhis2 context. `ui-forms` is meant to be nothing but a bridge between `ui-core` and RFF, so I'd either call this one `dhis2Password` or make it accept either a config object to be able to adjust these values or accept a pattern (which would then be a duplicate of the `createEqualTo` validator and therefor unnecessary)

File: src/validators/username.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import i18n from '@dhis2/d2-i18n'
+import { isEmpty, isString } from './helpers/index.js'
+
+const invalidUsernameMessage = i18n.t(
+    'Please provide a username between 1 and 255 characters'
+)
+
+const username = value =>
```
Comment (Mohammer5) [inline src/validators/username.js]: Same problem as `password`

**Review by Mohammer5 — CHANGES_REQUESTED** [review]
This is VERY nice! Will make working with validation a lot easier! Just a few small thing

File: src/validators/alphaNumeric.js (hunk 1 above)
Comment (HendrikThePendric) [reply src/validators/alphaNumeric.js]: OK, agreed and fixed in https://github.com/dhis2/ui-forms/pull/164/commits/98990f43d6258af21b8184e8bf206a30b8ef8d06

**Comment by ghost** [comment]
I think the one thing that I worry about is composing type validators with the required validator, related to what we discussed in this issue: https://github.com/dhis2/ui-forms/issues/151

It creates a situation where there's an interaction between two generic validators and managing the interaction seems more complicated to me than creating separate validators (so say requiredNumber and number). When composing we have to find a generic answer to what consitutes a value for `required` for example. Instead of just being able to define it per type.

I think in general, my preference would still be with singular validators (requiredNumber, number, requiredEmail, email), instead of focusing on composable ones. There might be more duplication, but it seems simpler to me. Easier to unit test as well as a validator just has a single task, instead of a task in multiple contexts. But there's been a lot of activity on this PR, and I might have missed discussion or misinterpreted something, so please correct me if I have!

**Comment by ghost** [comment]
All right, so @Mohammer5 and I talked a bit about the validators, and about the ones we expect to need. So composition seems necessary because we're expecting a lot of different types of validation to be necessary eventually. So that definitely makes sense 👍

I think the core concern for me was the `required` validator. I can imagine that it's slightly tricky for a dev implementing ui-core that the checks that required does are maybe not what you'd expect. Jan-Gerke suggested `notEmpty` as a name, which already seems like an improvement to me over required. Maybe it's not as big of an issue as I'm imagining it to be. But given that required does some very specific checks, maybe it'll help us scope what it does by naming it very specific. That'll communicate its function clearly to the user as well.

File: src/validators/createEqualTo.js (hunk 2 above)
Comment (HendrikThePendric) [reply src/validators/createEqualTo.js]: I think it's good you flagged this one up, but in https://github.com/dhis2/ui-forms/pull/164/commits/1d1094d840796f268c60a726ce963bb1dfaeff40 I have implemented it slightly differently than you suggested. I've kept the `isEmpty` check on the field itself, because:
- Theoretically both fields can have a different value when "being empty". One could be an empty string, and the other be `null`
- Regardless of how likely this is too happen, and if we should allow this to happen, I would like to keep all validators consistent, in the sense that they return `undefined` when the field is empty. And I want the same definition of "being empty" to be used everywhere.

If you agree, please resolve this conversation.

File: src/validators/createMaxCharacterLength.js (hunk 3 above)
Comment (HendrikThePendric) [reply src/validators/createMaxCharacterLength.js]: 🤦‍♂ oh dear.... good find. I made this one and the createMaxNumber at the same time. Obviously a string cannot have a negative length.... Fixed in https://github.com/dhis2/ui-forms/pull/164/commits/0c6a693b55e243b30e556fcde36a1127b4b5ef44

File: src/validators/password.js (hunk 5 above)
Comment (HendrikThePendric) [reply src/validators/password.js]: I've renamed to dhis2Password in https://github.com/dhis2/ui-forms/pull/164/commits/2d2685b0b49a18d0b4d71be1cd8ac025ae3e4e23

File: src/validators/username.js (hunk 6 above)
Comment (HendrikThePendric) [reply src/validators/username.js]: I've renamed to dhis2Username in https://github.com/dhis2/ui-forms/pull/164/commits/c00ada596d1c1119e79ac5162204d4067a36dcc7

File: src/validators/internationalPhoneNumber.js (hunk 4 above)
Comment (HendrikThePendric) [reply src/validators/internationalPhoneNumber.js:28]: > I think a validator is the wrong solution to this problem, if we want to have #.164 compliant phone number. We actually need an international phone number input component, which has an input or select for the country code and a second input for the number which trims leading zeros and special chars and then checks for a max length of 12.

This part I agree with, in principle. I definitely think that _in the future_ we should get a component like this and once it has been implemented we can probably get rid of this validator. I've have already done some research into this, and there are some excellent external libs that could power this component (i.e. [libphonenumber-js](https://github.com/catamphetamine/libphonenumber-js) or [Google's original libphonenumber](https://github.com/google/libphonenumber)). And the component itself could be a bit like [react-phone-number-input](https://github.com/catamphetamine/react-phone-number-input).

However, while I agree with the above _in principle_, I think that _in practice_ at the moment this validator does add some value. Mainly because we still have a hell of a lot of components to build and I can't imagine that this one would be very high on the priority list. So having a validator in place that validates as little as 80% of cases correctly will be a helpful thing to have.

And this also brings me to the other part of you comment:
> So the current implementation would return undefined for this: +49 0165 123456, although it doesn't match the E.164 standard, so this validator does not implement the E.164 correctly.

I am not claiming that the current algorithm is perfect by a long shot. It only does a very basic check of assessing whether a given value _could be_ a valid phone number:
- It removes characters that might be used a separators (hyphens, dots, spaces)
- It removes leading plus and zero characters
- It check if the cleaned string that is left represents a numerical value
- And if that string does not exceed 15 characters

My goal for this validator was simply to a) reject values that are definitely not valid b) not throw any false positives, i.e. don't reject correct phone numbers.

So, I do think that this validator adds _some_ value. But on the other hand, I can see why perhaps we don't want to add validators that are _roughly right_.

Starting this discussion is a bit of a slippery slope, because to a lesser extend, this would also apply to `email` and `url`. These probably have a much lower likelihood of accepting incorrect values, but they also don't cover 100% of cases.

Looking forward to finding out what you think we should do.

Comment (HendrikThePendric) [reply src/validators/internationalPhoneNumber.js:28]: BTW: why is `+49 0165 123456` invalid?

Comment (Mohammer5) [reply src/validators/internationalPhoneNumber.js:28]: > BTW: why is +49 0165 123456 invalid?
It is valid.. My bad

<hr />

> Starting this discussion is a bit of a slippery slope, because to a lesser extend, this would also apply to email and url. These probably have a much lower likelihood of accepting incorrect values, but they also don't cover 100% of cases.

@HendrikThePendric and I just quickly talked about this and came to the conclusion: The current way in this PR is fine, most important is: We don't show a warning when a value is correct and we catch most of the cases that are wrong

**Review by Mohammer5 — APPROVED** [review]
